File: backtracking/rec_permute/rec_permute.cpp
// string::substr
#include <iostream>
#include <string>

void RecPermute(std::string soFar, std::string rest){
    if (rest.empty()) {
        std::cout << soFar << std::endl;
    } else {
        for (int i=0; i < rest.length(); i++) {
            std::string remaining = rest.substr(0,i) + rest.substr(i+1);
            RecPermute(soFar + rest[i], remaining);
        }
    }
}


int main ()
{
    std::string a1 = "Good";
    std::string a2 = "";
    RecPermute(a2, a1);
    return 0;
}

File: backtracking/rec_permute/rec_permute.py
"""rec_permute.py

use the python3 to run this code
python3 rec_permute.py
"""


def rec_permute(so_far, rest):
    """This method finds the possible combinations for given string

    :param so_far: possible combinations of string.
    :type so_far: str
    :param rest: specify the actual string.
    :type rest: str
    :returns: None
    :raises: None
    """
    if not rest:
        print(so_far)
    else:
        for i in range(len(rest)):
            remain = rest[0:i] + rest[i+1:]
            rec_permute(so_far + rest[i], remain)


def main():
    """Main function
    """
    a = 'Hello'
    rec_permute("", a)


if __name__ == '__main__':
    main()
